File: server/tests/fixtures/alpha_rule_triggers/python/concurrency_lock_not_released.py
# Should trigger: concurrency.lock_not_released
import threading

lock = threading.Lock()

def critical_section():
    lock.acquire()
    if condition():
        return  # Early return without releasing lock!
    do_work()
    lock.release()
    
def condition():
    return True

def do_work():
    pass

File: server/tests/fixtures/alpha_rule_triggers/csharp/concurrency_lock_not_released.cs
// Should trigger: concurrency.lock_not_released
using System.Threading;

public class LockNotReleased {
    private readonly object _lock = new object();
    
    public void Process(bool condition) {
        Monitor.Enter(_lock);  // Not in try-finally!
        if (condition) {
            return;  // Early return without Exit!
        }
        DoWork();
        Monitor.Exit(_lock);
    }
    
    private void DoWork() {}
}
